src/main/scala/Service.scala

```diff
@@ -0,0 +1,160 @@
+  * The similar task example https://github.com/LvivScalaClub/cats-playground/blob/master/src/main/scala/BookRepository.scala
+  */
+
+case class User(id: Long, username: String)
+
+case class IotDevice(id: Long, userId: Long, sn: String)
+
+// NOTE: This import bring into the scope implicits that allow you to call .map and .flatMap on the type F[_]
+// and also bring you typeclasses that know how to flatmap (Monad) and map (Functor) over your higher-kinded type.
+import cats.implicits._
+
+trait Repository {
+  private var sequence = 0L
+
+  def nextId(): Long = {
+    sequence = sequence + 1
+    sequence
+  }
+}
+
+trait UserRepository[F[_]] extends Repository {
+  def registerUser(username: String): F[User]
+
+  def getById(id: Long): F[Option[User]]
+
+  def getByUsername(username: String): F[Option[User]]
+}
+
+class UserRepositoryId extends UserRepository[Id] {
+  private var storage: Map[Long, User] = Map()
+
+  override def registerUser(username: String): Id[User] = {
+    val id = nextId()
+    val user = User(id, username)
+    storage = storage + (id -> user)
+    user
+  }
+
+  override def getById(id: Long): Id[Option[User]] = storage.get(id)
+
+  override def getByUsername(username: String): Id[Option[User]] = storage.values.find(user => user.username == username)
+}
+
+class UserRepositoryFuture extends UserRepository[Future] {
+  private var storage: Map[Long, User] = Map()
+
+  override def registerUser(username: String): Future[User] = {
+    val id = nextId()
+    val user = User(id, username)
+    storage = storage + (id -> user)
+    Future.successful(user)
+  }
+
+  override def getById(id: Long): Future[Option[User]] = Future.successful(storage.get(id))
+
+  override def getByUsername(username: String): Future[Option[User]] =
+    Future.successful(storage.values.find(user => user.username == username))
+}
+
+trait IotDeviceRepository[F[_]] extends Repository {
+  def registerDevice(userId: Long, serialNumber: String): F[IotDevice]
+
+  def getById(id: Long): F[Option[IotDevice]]
+
+  def getBySn(sn: String): F[Option[IotDevice]]
+
+  def getByUser(userId: Long): F[Seq[IotDevice]]
+}
+
+class IotDeviceRepositoryId extends IotDeviceRepository[Id] {
+  private var storage: Map[Long, IotDevice] = Map()
+
+  override def registerDevice(userId: Long, serialNumber: String): Id[IotDevice] = {
+    val id = nextId()
+    val iotDevice = IotDevice(id, userId, serialNumber)
+    storage = storage + (id -> iotDevice)
+    iotDevice
+  }
+
+  override def getById(id: Long): Id[Option[IotDevice]] = storage.get(id)
+
+  override def getBySn(sn: String): Id[Option[IotDevice]] = storage.values.find(device => device.sn == sn)
+
+  override def getByUser(userId: Long): Id[Seq[IotDevice]] =
+    storage.values.filter(device => device.userId == userId).toSeq
+}
+
+class IotDeviceRepositoryFuture extends IotDeviceRepository[Future] {
```

Comment:
it is considered a good practice to have execution context as implicit param to that class so you can control later which EC executes your futures. Global EC should be used very carefully.